src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -535,16 +550,21 @@ event::ptr primitive_inst::realloc_if_needed() {
     }
 
     auto current_shape = updated_layout.get_shape();
-    auto& sp = *get_network().get_shape_predictor();
-    auto dt_size = ov::element::Type(updated_layout.data_type).bitwidth();
-    auto prealloc_info = sp.predict_preallocation_shape(id(), current_shape, dt_size, can_reuse_buffer);
+    std::pair<bool, ov::Shape> prealloc_info;
+    int32_t tmp_prealloc_count = _node->is_type<kv_cache>() ? kv_cache_inst::get_prealloc_iter_num() : -1;
+    GPU_DEBUG_IF(debug_config->mem_preallocation_params.is_initialized) {
+        // If debug config is set, repsect the config most
+        tmp_prealloc_count = -1;
+    }
+    prealloc_info = sp.predict_preallocation_shape(id(), current_shape, dt_size, can_reuse_buffer, tmp_prealloc_count);
```

Comment:
Here you add updated (padded) shape into shape predictor, but on the line 483 original shape is used. So won't it cause any issues with shape prediction?

Reply:
That's correct for one iteration only, right? But shape predictor is supposed to track some history, so I wonder if if will have completely different shapes due to that?

Reply:
Why one iteration only? allocate_ouput is called only when pad is 0 for kv cache with current impl. 
i.e., everytime it is calculating the shape w/o padding for kv cache.

Reply:
I checked shapes in those 2 places (1 - line 483, 2 - here) for multiple iterations using `KVCacheTests.smoke_multipleIterations_stateful_gather_with_initializer_batch_3` test:
```
kvcache:Concat_81 predict_preallocation_shape(2): [3,32,43,10]
kvcache:Concat_81 predict_preallocation_shape(1): [3,32,21,10]
kvcache:Concat_81 predict_preallocation_shape(2): [3,32,44,10]
kvcache:Concat_81 predict_preallocation_shape(1): [3,32,23,10]
kvcache:Concat_81 predict_preallocation_shape(2): [3,32,49,10]
kvcache:Concat_81 predict_preallocation_shape(1): [3,32,25,10]
kvcache:Concat_81 predict_preallocation_shape(2): [3,32,45,10]
kvcache:Concat_81 predict_preallocation_shape(1): [3,32,27,10]
kvcache:Concat_81 predict_preallocation_shape(2): [3,32,47,10]
kvcache:Concat_81 predict_preallocation_shape(1): [3,32,29,10]
kvcache:Concat_81 predict_preallocation_shape(2): [3,32,49,10]
```

As you can see, sequence value is different and we have always switches between larger and smaller shape. So I'm curious how that impacts shape predictor work. 

Anyway, that doesn't look like a critical issue, so we can merge it as is and investigate it more later.

Reply:
Thanks for checking. I will check with that behaviour..